pkg/proxy/winkernel/hcnutils.go

```diff
@@ -346,13 +376,157 @@ func (hns hns) getLoadBalancer(endpoints []endpointsInfo, flags loadBalancerFlag
+	return hcn.ListEndpoints()
+}
+
+func (h *RealHCN) GetEndpointByName(endpointName string) (*hcn.HostComputeEndpoint, error) {
+	return hcn.GetEndpointByName(endpointName)
+}
+
+func (h *RealHCN) ListLoadBalancers() ([]hcn.HostComputeLoadBalancer, error) {
+	return hcn.ListLoadBalancers()
+}
+
+func (h *RealHCN) GetLoadBalancerByID(loadBalancerId string) (*hcn.HostComputeLoadBalancer, error) {
+	return hcn.GetLoadBalancerByID(loadBalancerId)
+}
+
+func (h *RealHCN) CreateEndpoint(endpoint *hcn.HostComputeEndpoint, network *hcn.HostComputeNetwork) (*hcn.HostComputeEndpoint, error) {
+	return network.CreateEndpoint(endpoint)
+}
+
+func (h *RealHCN) CreateRemoteEndpoint(endpoint *hcn.HostComputeEndpoint, network *hcn.HostComputeNetwork) (*hcn.HostComputeEndpoint, error) {
+	return network.CreateRemoteEndpoint(endpoint)
+}
+
+func (h *RealHCN) CreateLoadBalancer(loadbalancer *hcn.HostComputeLoadBalancer) (*hcn.HostComputeLoadBalancer, error) {
+	return loadbalancer.Create()
+}
+
+func (h *RealHCN) DeleteLoadBalancer(loadbalancer *hcn.HostComputeLoadBalancer) error {
+	return loadbalancer.Delete()
+}
+
+func (h *RealHCN) DeleteEndpoint(endpoint *hcn.HostComputeEndpoint) error {
+	return endpoint.Delete()
+}
+
+// Here are some helper methods for proxier.go
+
+func getNetworkName(hnsNetworkName string) (string, error) {
+	if len(hnsNetworkName) == 0 {
+		klog.V(3).InfoS("Flag --network-name not set, checking environment variable")
+		hnsNetworkName = os.Getenv("KUBE_NETWORK")
+		if len(hnsNetworkName) == 0 {
+			return "", fmt.Errorf("Environment variable KUBE_NETWORK and network-flag not initialized")
+		}
+	}
+	return hnsNetworkName, nil
+}
+
+func getNetworkInfo(hns IHCNUtils, hnsNetworkName string) (*hnsNetworkInfo, error) {
+	hnsNetworkInfo, err := hns.getNetworkByName(hnsNetworkName)
+	for err != nil {
+		klog.ErrorS(err, "Unable to find HNS Network specified, please check network name and CNI deployment", "hnsNetworkName", hnsNetworkName)
+		time.Sleep(1 * time.Second)
+		hnsNetworkInfo, err = hns.getNetworkByName(hnsNetworkName)
+	}
+	return hnsNetworkInfo, err
+}
+
+func isOverlay(hnsNetworkInfo *hnsNetworkInfo) bool {
+	return strings.EqualFold(hnsNetworkInfo.networkType, NETWORK_TYPE_OVERLAY)
+}
+
+func (svcInfo *serviceInfo) cleanupAllPolicies(endpoints []proxy.Endpoint) {
+	klog.V(3).InfoS("Service cleanup", "serviceInfo", svcInfo)
+	// Skip the svcInfo.policyApplied check to remove all the policies
+	svcInfo.deleteAllHnsLoadBalancerPolicy()
+	// Cleanup Endpoints references
+	for _, ep := range endpoints {
+		epInfo, ok := ep.(*endpointsInfo)
+		if ok {
+			epInfo.Cleanup()
+		}
+	}
+	if svcInfo.remoteEndpoint != nil {
+		svcInfo.remoteEndpoint.Cleanup()
+	}
+
+	svcInfo.policyApplied = false
+}
+
+func (svcInfo *serviceInfo) deleteAllHnsLoadBalancerPolicy() {
+	// Remove the Hns Policy corresponding to this service
+	hns := svcInfo.hns
+	hns.deleteLoadBalancer(svcInfo.hnsID)
```

Comment:
this may return an error. I wonder if it's ok for us to simply ignore them, or we should at least log the errors. Same as the calls below. Asking this because in ``deleteAllHnsLoadBalancerPolicy`` below, we actually log an error if the loadbalancer fails to delete, so at least the behaviour is a bit inconsistent.